### PR TITLE
HDDS-8777. Ratis Echo Freon Command

### PR DESCRIPTION
**Pull request by tanvipenumudy**
## What changes were proposed in this pull request?

To implement an ozone freon extension that sends a request solely used to benchmark Ratis operations per second for log append.

- Write to Ratis log.
- During applyTransaction, do not write to RocksDB but complete the request ASAP.

This shall help in profiling Ratis' performance within a customer's environment and help debug performance issues. This shall also help establish an upper bound for OM's write performance.

Added an optional `--ratis` flag to `ozone freon om-echo` for writing to Ratis log (flag to be skipped for read-only echo RPC request to the OM - with/without payload).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8777

## How was this patch tested?

- Added robot tests. 
- Tested manually using the `ozone debug ratislogparser` command on a cluster with the basic Ozone services running:

```
[Before] ozone debug ratislogparser om -s=/var/lib/hadoop-ozone/om/ratis/path/to/segment/file | grep -i EchoRPC | wc -l
Output: 0

Command: ozone freon om-echo --ratis

[After] ozone debug ratislogparser om -s=/var/lib/hadoop-ozone/om/ratis/path/to/segment/file | grep -i EchoRPC | wc -l
Output: 1000
```
- Green Git CI/CD: https://github.com/tanvipenumudy/ozone/actions/runs/6070498943

### CODE REVIEW

**Comment by tanvipenumudy** [comment]
Could you please take a look @kerneltime, @duongkame, @sadanand48? Thanks!

**Comment by kerneltime** [comment]
The change looks good, we should send the payload in response if configured.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/util/OMEchoRPCWriteRequest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+
+    OMClientResponse omClientResponse = null;
+    OzoneManagerProtocolProtos.OMResponse.Builder omResponse =
+        OmResponseUtil.getOMResponseBuilder(getOmRequest());
+    omClientResponse = new OMEchoRPCWriteResponse(omResponse.build());
```
[the remark this answers is not in the document]
Comment (kerneltime) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/util/OMEchoRPCWriteRequest.java:44]: +1

[the remark this answers is not in the document]
Comment (tanvipenumudy) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/util/OMEchoRPCWriteRequest.java]: Thank you, made the required changes!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/util/OMEchoRPCWriteRequest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+
+    EchoRPCRequest echoRPCRequest = getOmRequest().getEchoRPCRequest();
+
+    byte[] payloadBytes = new byte[0];
+    int payloadRespSize = Math.min(
+        echoRPCRequest.getPayloadSizeResp() * RPC_PAYLOAD_MULTIPLICATION_FACTOR,
+        MAX_SIZE_KB);
+    if (payloadRespSize > 0) {
+      payloadBytes = RandomUtils.nextBytes(payloadRespSize);
```
[the remark this answers is not in the document]
Comment (tanvipenumudy) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/util/OMEchoRPCWriteRequest.java]: Thanks, made the changes!

**Comment by adoroszlai** [comment]
@tanvipenumudy `Transfer Leadership for OM` is consistently timing out with yor patch, please check.  To reproduce it, you can trim `ozonesecure-ha/test.sh` to:

```
execute_robot_test s3g freon
execute_robot_test s3g omha/om-leader-transfer.robot
```

Without `freon` leadership transfer test passes, so I assume it is caused by the updated echo request.

**Comment by tanvipenumudy** [comment]
Thank you, @adoroszlai. I will reproduce and address the timeout issue as suggested.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/PayloadUtils.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  public static byte[] generatePayloadBytes(int payloadSize) {
+
+    byte[] payloadBytes = new byte[0];
+    int payloadRespSize =
+        Math.min(payloadSize * RPC_PAYLOAD_MULTIPLICATION_FACTOR, MAX_SIZE_KB);
+    if (payloadRespSize > 0) {
+      payloadBytes = RandomUtils.nextBytes(payloadRespSize);
+    }
+
+    return payloadBytes;
```
Comment (duongkame) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/PayloadUtils.java:42]:
```suggestion
    int payloadRespSize =
        Math.min(payloadSize * RPC_PAYLOAD_MULTIPLICATION_FACTOR, MAX_SIZE_KB);
    if (payloadRespSize > 0) {
      return RandomUtils.nextBytes(payloadRespSize);
    }
```

**Comment by adoroszlai** [comment]
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/PayloadUtils.java:[40,3] missing return statement
```

https://github.com/tanvipenumudy/ozone/actions/runs/5344335030/jobs/9688599409#step:6:2040

**Review by duongkame — CHANGES_REQUESTED** [review]
Hi @tanvipenumudy, please refer to [this change](https://github.com/duongkame/ozone/commit/6f117de2edb27cf801f9fbad9c0001996f98779a) to fix the test failure. 

It turns out that every ratis write has to pass the transaction index to the double buffer, so that the proper last transaction index is updated to the transaction metadata table. Without the right tnx index updated in the tnx metadata table, the new leader after leadership transfer doesn’t see the right latest log idx and thus fails to process new log entries.

**Comment by tanvipenumudy** [comment]
Thank you @duongkame for the review, I have made the required changes.

**Review by duongkame — APPROVED** [review]
Thanks for the changes, @tanvipenumudy. LGTM.

**Pull request merged**